### PR TITLE
[#11572] Notification Feature - Add tests for notification POST route

### PR DESCRIPTION
**Pull request by NicolasCwy**
Part of  #11572

> To be merged after PR #11648 

**Outline of Solution**
- Add utility getNotification method to `LogicExtension` (`getNotification()` not used in API and only for testing)
- Test logic `getNotification(String)` method
- [X] Add test for `CreateNotificationAction`
- [x] Add POST test for logic layer

NOTE: Change startTimestamp and endTimestamp type from Long to long primitive in `NotificationData` and `CreateNotificationRequest`.

### CODE REVIEW

File: src/test/java/teammates/common/datatransfer/attributes/NotificationAttributesTest.java
Comment (t-cheepeng) [inline src/test/java/teammates/common/datatransfer/attributes/NotificationAttributesTest.java:194]: This makes the test seem like you are testing `getCopy`. Probably should build another attribute object that is exactly the same and do the `equals`.

File: src/test/java/teammates/common/datatransfer/attributes/NotificationAttributesTest.java
Comment (t-cheepeng) [inline src/test/java/teammates/common/datatransfer/attributes/NotificationAttributesTest.java:224]: `null` value test?

File: src/test/java/teammates/logic/core/NotificationsLogicTest.java
Comment (t-cheepeng) [inline src/test/java/teammates/logic/core/NotificationsLogicTest.java:94]: Aren't two notificationattributes equal if id is the same? Why the need for this assertEquals

File: src/test/java/teammates/test/BaseTestCaseWithDatabaseAccess.java
Comment (t-cheepeng) [inline src/test/java/teammates/test/BaseTestCaseWithDatabaseAccess.java:79]: Unncessary line change

File: src/test/java/teammates/ui/webapi/GetNotificationActionTest.java
Comment (t-cheepeng) [inline src/test/java/teammates/ui/webapi/GetNotificationActionTest.java:164]: you mean `usertype is a invalid string`?

**Review by t-cheepeng — CHANGES_REQUESTED** [review]
Great coverage!

File: src/main/java/teammates/ui/output/NotificationData.java
Comment (NicolasCwy) [reply src/main/java/teammates/ui/output/NotificationData.java:40]: I made this change because I changed the input from the user is changed to `long` in `CreateNotificationRequest.java` as there was no reason to take a `Long`, I can refactor `createdAt` and `updatedAt` if I find there are no dependencies

File: src/main/java/teammates/logic/api/Logic.java
Comment (NicolasCwy) [reply src/main/java/teammates/logic/api/Logic.java:99]: Preconditions in the logic API seemed to be marked with an extra *, possibly to highlight this

File: src/test/java/teammates/logic/core/NotificationsLogicTest.java
Comment (NicolasCwy) [reply src/test/java/teammates/logic/core/NotificationsLogicTest.java:94]: Since I was not able to use the parent classes `verifyPresentInDatabase`, I was trying to verify that the database contained the notification requested by the user, therefore every single field should be the same. This was what I had proposed during our meeting

File: src/test/java/teammates/test/BaseTestCaseWithDatabaseAccess.java
Comment (NicolasCwy) [reply src/test/java/teammates/test/BaseTestCaseWithDatabaseAccess.java:79]: Will amend this in the next commit

File: src/test/java/teammates/logic/core/NotificationsLogicTest.java
Comment (t-cheepeng) [reply src/test/java/teammates/logic/core/NotificationsLogicTest.java:94]: I see. I thought that only applied to the logic method calls. Just add a comment explaining why the need for such a method 👍

File: src/main/java/teammates/ui/output/NotificationData.java
Comment (t-cheepeng) [reply src/main/java/teammates/ui/output/NotificationData.java:40]: Please update for the created/updated too

File: src/test/java/teammates/ui/webapi/CreateNotificationActionTest.java
Comment (jianhandev) [inline src/test/java/teammates/ui/webapi/CreateNotificationActionTest.java:73]: Since you are calling this method over and over again and only vary one parameter at a time, it's better to just create a typical request using `getTypicalCreateRequest()` and then set attribute for the request object to `null`.

**Review by jianhandev — CHANGES_REQUESTED** [review]

File: src/main/java/teammates/storage/api/NotificationsDb.java
Comment (NicolasCwy) [reply src/main/java/teammates/storage/api/NotificationsDb.java:109]: I have mistakenly merged it into this PR, perhaps we can merge this PR after PR#11651

File: src/test/java/teammates/ui/webapi/CreateNotificationActionTest.java
Comment (NicolasCwy) [reply src/test/java/teammates/ui/webapi/CreateNotificationActionTest.java:73]: Would that reduce the readability of the code since I would not be able to put it concisely in a single line within  `verifyHttpRequestBodyFailure`, understand that this refactor would be better if we increase the number of test cases.

Comment (jianhandev) [reply src/test/java/teammates/ui/webapi/CreateNotificationActionTest.java:73]: I would actually think readability will improve: 
```
request = getTypicalCreateRequest();
request.setTitle(null);
verifyHttpRequestBodyFailure(request);
```

Notice that now you don't have long lines of code, and it is much easier to tell that you are testing for `null` title

File: src/test/java/teammates/logic/core/NotificationsLogicTest.java
Comment (jianhandev) [inline src/test/java/teammates/logic/core/NotificationsLogicTest.java:36]: Should add tests for `updateNotification`, `deleteNotification`, `getAllNotifications` and `getActiveNotificationsByTargetUser` as well, you can make use of test data from typical data bundle.

Comment (NicolasCwy) [reply src/test/java/teammates/logic/core/NotificationsLogicTest.java:36]: I was originally thinking of splitting it into individual PRs for each route so that [PR#11665](https://github.com/TEAMMATES/teammates/pull/1166) can be merged ASAP. Since PR#11665 makes some changes to POST and requires these tests to ensure its correctness.

Comment (jianhandev) [reply src/test/java/teammates/logic/core/NotificationsLogicTest.java:36]: I see, but #11665 seems to be blocked by Notification feature - Database Test #11651 which has pending changes, maybe consider adding `updateNotification` and `deleteNotification` first?

Comment (NicolasCwy) [reply src/test/java/teammates/logic/core/NotificationsLogicTest.java:36]: Sure thing, does the POST tests look fine as of right now?

Comment (jianhandev) [reply src/test/java/teammates/logic/core/NotificationsLogicTest.java:36]: The rest of this PR looks pretty ok to me, except the `verifyPresentInDatabase` which still kind of bugs me.

Comment (NicolasCwy) [reply src/test/java/teammates/logic/core/NotificationsLogicTest.java:36]: Agreed, this can be discussed, as the only way I can think of is creating the route for `getNotification(notifId)` purely for testing purposes

Comment (jianhandev) [reply src/test/java/teammates/logic/core/NotificationsLogicTest.java:36]: @zhaojj2209 yes I think it is reasonable too, take note that the current resource URI for the GET endpoint (`ResourceURIs.NOTIFICATION`) will need to be changed to `ResourceURIs.NOTIFICATIONS`. `ResourceURIs.NOTIFICATION` should be for getting single notification.

**Review by jianhandev — APPROVED** [review]

**Pull request merged**
